astropy/coordinates/builtin_frames/astrometric.py

```diff
@@ -0,0 +1,155 @@
+    origin_dec : `Angle`, optional, must be keyword
+        The Declination (Dec) of the Galactic center in the ICRS frame.
+
+    Examples
+    --------
+    To transform to the Astrometric frame with the default
+    frame attributes, pass the uninstantiated class name to the
+    ``transform_to()`` method of a coordinate frame or
+    `~astropy.coordinates.SkyCoord` object::
+
+        >>> import astropy.units as u
+        >>> import astropy.coordinates as coord
+        >>> c = coord.ICRS(ra=[158.3122, 24.5] * u.degree,
+        ...                dec=[-17.3, 81.52] * u.degree,
+        ...                distance=[11.5, 24.12] * u.kpc)
+        >>> c.transform_to(coord.Astrometric) # doctest: +FLOAT_CMP
+        <Astrometric Coordinate (origin_distance=8.3 kpc, origin_ra=266d24m18.36s, origin_dec=-28d56m10.23s, z_sun=27.0 pc, roll=0.0 deg): (x, y, z) in kpc
+            [(-9.6083818980977, -9.400621883358546, 6.520560663896347),
+             (-21.283023068029138, 18.763340128812384, 7.846938548636718)]>
+
+    To specify a custom set of parameters, you have to include extra keyword
+    arguments when initializing the Astrometric frame object::
+
+        >>> c.transform_to(coord.Astrometric(origin_distance=8.1*u.kpc)) # doctest: +FLOAT_CMP
+        <Astrometric Coordinate (origin_distance=8.1 kpc, origin_ra=266d24m18.36s, origin_dec=-28d56m10.23s, z_sun=27.0 pc, roll=0.0 deg): (x, y, z) in kpc
+            [(-9.407859235565343, -9.400621883358546, 6.520665737962164),
+             (-21.08239383088295, 18.763340128812384, 7.84798134569032)]>
+
+    Similarly, transforming from the Astrometric frame to another coordinate frame::
+
+        >>> c = coord.Astrometric(x=[-8.3, 4.5] * u.kpc,
+        ...                          y=[0., 81.52] * u.kpc,
+        ...                          z=[0.027, 24.12] * u.kpc)
+        >>> c.transform_to(coord.ICRS) # doctest: +FLOAT_CMP
+        <ICRS Coordinate: (ra, dec, distance) in (deg, deg, kpc)
+            [(86.22349058727241, 28.8389413808627, 4.391577882957292e-05),
+             (289.6680265194508, 49.88763881149547, 85.96407345372828)]>
+
+    Or, with custom specification of the Galactic center::
+
+        >>> c = coord.Astrometric(x=[-8.0, 4.5] * u.kpc,
+        ...                          y=[0., 81.52] * u.kpc,
+        ...                          z=[21.0, 24120.0] * u.pc,
+        ...                          z_sun=21 * u.pc, origin_distance=8. * u.kpc)
+        >>> c.transform_to(coord.ICRS) # doctest: +FLOAT_CMP
+        <ICRS Coordinate: (ra, dec, distance) in (deg, deg, kpc)
+            [(86.25852490164378, 28.85773187391088, 2.7562547481200286e-05),
+             (289.77285254989323, 50.062904565432014, 85.92160096237191)]>
+
+    """
+    default_representation = CartesianRepresentation
+
+    origin_ra = QuantityFrameAttribute(default=0, unit=u.degree)
+    origin_dec = QuantityFrameAttribute(default=0, unit=u.degree)
```

Comment:
Could you use a custom frame attribute here which holds a SkyCoord? That way you could convert the origin to the reference frame (ICRS) automatically for users.

I did this like this:
```python
class CoordinateLocationAttribute(FrameAttribute):
    """A frame attribute which is a coordinates object."""
    
    def convert_input(self, value):
        """
        Checks that the input is a SkyCoord with the necessary units (or the
        special value ``None``).

        Parameters
        ----------
        value : object
            Input value to be converted.

        Returns
        -------
        out, converted : correctly-typed object, boolean
            Tuple consisting of the correctly-typed object and a boolean which
            indicates if conversion was actually performed.

        Raises
        ------
        ValueError
            If the input is not valid for this attribute.
        """
        if value is None:
            return None, False
        elif isinstance(value, ICRS):
            return value, False
        else:
            if not hasattr(value, 'transform_to'):
                raise ValueError('"{0}" was passed into an '
                                 'CoordinateLocationAttribute, but it does not have '
                                 '"transform_to" method'.format(value))
            icrsobj = value.transform_to(ICRS)
            if isinstance(icrsobj, SkyCoord):
                return icrsobj.frame, True
            return icrsobj, True
```

and then replacing these lines with
```python
origin = CoordinateLocationAttribute()
```

I think this would also make it easier to factor out the base reference frame, if we can figure out how to make the transforms work for arbitrary target frames.